pkg/sources/forward/data_forward.go

```diff
@@ -192,6 +192,31 @@ type readWriteMessagePair struct {
 	transformerError error
 }
 
+func processReadMessages(hasUDTransformer bool, readMessages []*isb.ReadMessage, processorWM wmb.Watermark,
+	transformerResults []readWriteMessagePair, df *DataForward, transformerCh chan *readWriteMessagePair) {
+	for idx, m := range readMessages {
+		// emit message size metric
+		metrics.ReadBytesCount.With(map[string]string{
+			metrics.LabelVertex:             df.vertexName,
+			metrics.LabelPipeline:           df.pipelineName,
+			metrics.LabelVertexType:         string(dfv1.VertexTypeSource),
+			metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)),
+			metrics.LabelPartitionName:      df.reader.GetName(),
+		}).Add(float64(len(m.Payload)))
+		// assign watermark to the message
+		m.Watermark = time.Time(processorWM)
+		transformerResults[idx].readMessage = m
+
+		if hasUDTransformer {
+			// send transformer processing work to the channel
+			transformerCh <- &transformerResults[idx]
+		} else {
+			// if no user-defined transformer exists, then the messages to write will be identical to the messages read from source
+			transformerResults[idx].writeMessages = []*isb.WriteMessage{{Message: m.Message}}
```

Comment:
* With this change, is transformerResults still a concise name?
* It looks we are assigning writeMessages within processReadMessages. Does `processReadMessages` own populating writeMessages? If so, should it also populate when a transformer exists?

We can sync up on this tmr morning in the office.